### PR TITLE
Add Optional InfiniteList Support to AgendaList

### PR DESCRIPTION
**Pull request by yakirza17**
Introduces an experimental prop `infiniteListProps` to `AgendaList`. When defined, `AgendaList` switches from `SectionList` to `InfiniteList`. Be aware, as an experimental feature, it might not support all existing APIs.

Additionally, enhances `InfiniteList` to accept a `layoutProvider` for detailed item size control.

### CODE REVIEW

File: src/expandableCalendar/agendaListNew.tsx [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+      list.current?.scrollToIndex(sectionIndex, true);
+    }
+  }, 1000, {leading: false, trailing: true}), [ sections]);
+
+  const layoutProvider = useMemo(
+    () => new LayoutProvider(
+      (index) => data[index]?.isTitle ? 'title': 'page',
+      (type, dim) => {
+        dim.width = constants.screenWidth;
+        dim.height = type === 'title' ? 60 : 120;
```
Comment (yakirza17) [inline src/expandableCalendar/agendaListNew.tsx:147]: should see what the right size to pass

Comment (chenei) [reply src/expandableCalendar/agendaListNew.tsx:147]: @ethanshar @Inbal-Tish please give your thoughts

File: src/infinite-list/index.tsx [hunk 2]
```diff
@@ -157,13 +162,17 @@ const InfiniteList = (props: InfiniteListProps, ref: any) => {
...
       dataProvider={dataProvider}
-      layoutProvider={layoutProvider.current}
+      layoutProvider={props.layoutProvider ?? layoutProvider.current}
       extendedState={extendedState}
       initialRenderIndex={initialPageIndex}
       renderAheadOffset={5 * pageWidth}
       onScroll={onScroll}
       style={style}
       scrollViewProps={scrollViewPropsMemo}
+      onMomentumScrollEnd={scrollViewProps?.onMomentumScrollEnd}
```
Comment (Inbal-Tish) [inline src/infinite-list/index.tsx:172]: scrollViewPropsMemo should include 'onMomentumScrollEnd', why separate them?

File: src/infinite-list/index.tsx [hunk 3]
```diff
@@ -67,13 +68,17 @@ const InfiniteList = (props: InfiniteListProps, ref: any) => {
   const reloadPagesDebounce = useCallback(debounce(reloadPages, 500, {leading: false, trailing: true}), [reloadPages]);
 
   useEffect(() => {
+    if (props.onEndReached) {
```
Comment (Inbal-Tish) [inline src/infinite-list/index.tsx:71]: Why this change? Even if you pass 'onEndReached' callback you need to scroll the data in case it changes... this is not clear

File: src/expandableCalendar/agendaListNew.tsx [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+import Context from './Context';
+import constants from "../commons/constants";
+import {parseDate} from "../interface";
+import {LayoutProvider} from "recyclerlistview/dist/reactnative/core/dependencies/LayoutProvider";
+import {AgendaListProps, AgendaSectionHeader} from "./agendaList";
+
+/**
+ * @description: AgendaList component that use InfiniteList to improve performance
+ * @note: Should be wrapped with 'CalendarProvider'
+ * @extends: SectionList
```
Comment (Inbal-Tish) [inline src/expandableCalendar/agendaListNew.tsx:31]: Doesn't extend SectionList, but Infinite list

File: src/expandableCalendar/agendaListNew.tsx [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+ * @example: https://github.com/wix/react-native-calendars/blob/master/example/src/screens/expandableCalendar.js
+ */
+const AgendaList = (props: AgendaListProps) => {
```
Comment (Inbal-Tish) [inline src/expandableCalendar/agendaListNew.tsx:34]: Rename the component AgendaListNew or InfiniteAgendaList

File: src/expandableCalendar/agendaListNew.tsx [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    if (props.renderItem) {
+      return props.renderItem({item} as any);
+    }
+
+    return <></>;
+  }, [props.renderItem]);
+
+  const _onEndReached = useCallback(() => {
+    if (props.onEndReached) {
+      props.onEndReached({distanceFromEnd: 0});
```
Comment (Inbal-Tish) [inline src/expandableCalendar/agendaListNew.tsx:228]: Why distanceFromEnd is hard-coded to 0?

File: src/expandableCalendar/agendaListNew.tsx [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+  }, [data]);
+
+  const _onMomentumScrollEnd = useCallback(() => {
+    sectionScroll.current = false;
+  }, []);
+
+  const headerTextStyle = useMemo(() => [style.current.sectionText, sectionStyle], [sectionStyle]);
+
+  const onHeaderLayout = useCallback((event: LayoutChangeEvent) => {
+    sectionHeight.current = event.nativeEvent.layout.height;
```
Comment (Inbal-Tish) [inline src/expandableCalendar/agendaListNew.tsx:200]: Where are you using the sectionHeight.current?

File: src/expandableCalendar/agendaListNew.tsx [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+  }, []);
+
+  const scrollToSection = useCallback(debounce((d) => {
+    const sectionIndex = scrollToNextEvent ? getNextSectionIndex(d) : getSectionIndex(d);
+    if (isUndefined(sectionIndex)) {
+      return;
+    }
+
+    if (list?.current && sectionIndex !== undefined) {
+      sectionScroll.current = true; // to avoid setDate() in onViewableItemsChanged
```
Comment (Inbal-Tish) [inline src/expandableCalendar/agendaListNew.tsx:135]: You're not using onViewableItemsChanged here so why onViewableItemsChanged still set?

Comment (Inbal-Tish) [inline src/expandableCalendar/agendaListNew.tsx:135]:
```suggestion
      sectionScroll.current = true; // to avoid setDate() in _onVisibleIndicesChanged
```

File: src/expandableCalendar/agendaListNew.tsx (hunk 1 above)
Comment (Inbal-Tish) [reply src/expandableCalendar/agendaListNew.tsx:147]: In the current example, it should be 80 instead of 120, but anyways you'll have to let the user pass the item's height to set it there.

**Comment by Inbal-Tish** [comment]
@yakirza17 @chenei Overall looks good. Left a few comments

File: src/expandableCalendar/agendaListNew.tsx (hunk 7 above)
Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx:200]: removed

File: src/expandableCalendar/agendaListNew.tsx (hunk 6 above)
Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx:228]: The `onEndReached` of `RecyclerListView` is not pass any argument, so I don't have this information

File: src/expandableCalendar/agendaListNew.tsx (hunk 5 above)
Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx:34]: Renamed to `InfiniteAgendaList`

File: src/infinite-list/index.tsx (hunk 3 above)
Comment (yakirza17) [reply src/infinite-list/index.tsx:71]: I don't want to scroll in the case of `onEndReached` I want to keep the scroll position in the same place (without this change, the view jump to the first item on any array update)
In case of new data changes, we have `scrollToIndex` on `InfiniteAgendaList`

File: src/infinite-list/index.tsx (hunk 2 above)
Comment (yakirza17) [reply src/infinite-list/index.tsx:172]: removed

File: src/expandableCalendar/agendaListNew.tsx (hunk 1 above)
Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx:147]: Created `infiniteListProps` in the oldAgendaList, with an option to define these values (and load the new component in case it pass - even an empty object)

Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx:147]: @Inbal-Tish It's ok to put values like that? will it work in any view? tablet view? landscape?

File: src/infinite-list/index.tsx (hunk 3 above)
Comment (Inbal-Tish) [reply src/infinite-list/index.tsx:71]: You want to block the scroll when onEndReached is called or implemented?

Comment (yakirza17) [reply src/infinite-list/index.tsx:71]: If it exists

File: src/expandableCalendar/agendaListNew.tsx (hunk 6 above)
Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx]: It’s required on the onEndReach on the agenda list props, and I don’t want to change existing props

File: src/infinite-list/index.tsx (hunk 3 above)
Comment (Inbal-Tish) [reply src/infinite-list/index.tsx:71]: But if the user wants to change the screen's background color every time the onEndReach is called, he needs to implement it, it doesn't mean he doesn't want the scroll to happen...

**Comment by Inbal-Tish** [comment]
@yakirza17 Can you add a code snippet for the AgendaList so I can see on what example you're working on?

File: src/infinite-list/index.tsx (hunk 3 above)
Comment (yakirza17) [reply src/infinite-list/index.tsx:71]: Ok, I will add a new props on infinite List that will disable this scroll

Comment (yakirza17) [reply src/infinite-list/index.tsx:71]: BTW, the infinite List does not supports onEndReached before this PR

**Comment by yakirza17** [comment]
@Inbal-Tish, add `infiniteListProps={{}}` in the AgendaList on `example/src/screens/expandableCalendarScreen.tsx`
I also test it with our real-use repo (by changing the node modules code)

File: src/expandableCalendar/agendaListNew.tsx (hunk 1 above)
Comment (Inbal-Tish) [reply src/expandableCalendar/agendaListNew.tsx]: It will stay the same unless you change it when the orientation change. Regarding tablets, it depends on the design of course... Once you give the user the option to set these values it can be dynamic. 'pageWidth' and 'pageHeight' covers that

File: src/expandableCalendar/agendaListNew.tsx (hunk 6 above)
Comment (Inbal-Tish) [reply src/expandableCalendar/agendaListNew.tsx]: It a FlatList prop: onEndReached?: ((info: { distanceFromEnd: number }) => void) | null | undefined; So you can return undefined

Comment (yakirza17) [reply src/expandableCalendar/agendaListNew.tsx]: I want to apply this CB for users that want to know when the end is reached, for applying this CB I must provide an info parameter (it's not optional), the null | undefined part just says that it's not required CB

image

Comment (Inbal-Tish) [reply src/expandableCalendar/agendaListNew.tsx]: Ok... Add a comment about that please

File: src/infinite-list/index.tsx (hunk 3 above)
Comment (Inbal-Tish) [reply src/infinite-list/index.tsx]: Yes, I noticed

File: src/infinite-list/index.tsx [hunk 9]
```diff
@@ -67,13 +69,17 @@ const InfiniteList = (props: InfiniteListProps, ref: any) => {
   const reloadPagesDebounce = useCallback(debounce(reloadPages, 500, {leading: false, trailing: true}), [reloadPages]);
 
   useEffect(() => {
+    if (props.disableScrollOnDataChange) {
```
Comment (Inbal-Tish) [inline src/infinite-list/index.tsx:72]: instead of extracting using props.xxx you can add it above where we extract all props. Also for the prop you're passing to the RecyclerListView

File: src/infinite-list/index.tsx [hunk 10]
```diff
@@ -157,13 +164,16 @@ const InfiniteList = (props: InfiniteListProps, ref: any) => {
       isHorizontal={isHorizontal}
       rowRenderer={renderItem}
       dataProvider={dataProvider}
-      layoutProvider={layoutProvider.current}
+      layoutProvider={props.layoutProvider ?? layoutProvider.current}
```
Comment (Inbal-Tish) [inline src/infinite-list/index.tsx:167]: here too

File: src/infinite-list/index.tsx [hunk 11]
```diff
@@ -157,13 +164,16 @@ const InfiniteList = (props: InfiniteListProps, ref: any) => {
...
       extendedState={extendedState}
       initialRenderIndex={initialPageIndex}
       renderAheadOffset={5 * pageWidth}
       onScroll={onScroll}
       style={style}
       scrollViewProps={scrollViewPropsMemo}
+      onEndReached={props?.onEndReached}
```
Comment (Inbal-Tish) [inline src/infinite-list/index.tsx:174]: the 3 here too

**Review by Inbal-Tish — APPROVED** [review]
